File: Day32/swap_linkedlist_nodes.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:

        curr = head
        op = []

        while curr:
            op.append(curr.val)
            curr = curr.next

        opll = head

        for i in range(0,len(op)-1,2):
            op[i], op[i+1] = op[i+1], op[i]

        for i in range(len(op)):
            opll.val = op[i]
            opll = opll.next

        return head




            

File: Day32/swap_linkedlist_nodes.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* recurse(ListNode* head)
    {
        if(head == NULL || head->next == NULL)
        {
            return head;
        }

        //create a temp node and point it to next node
        ListNode* temp = head->next;

        //make recursive call to swap
        head->next = recurse(head->next->next);

        //now put temp's next node as head
        temp->next = head;

        return temp;
    }

    ListNode* swapPairs(ListNode* head) {

        return recurse(head);
    }
};
